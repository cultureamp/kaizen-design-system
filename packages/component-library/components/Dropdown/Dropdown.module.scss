@import "../../styles/color";
@import "../../styles/layers";
@import "../../styles/type";
@import "~@kaizen/design-tokens/sass/border";
@import "~@kaizen/design-tokens/sass/typography";
@import "~@kaizen/deprecated-component-library-helpers/styles/type";

$width: 248px;
$caButton-border-width: $kz-border-solid-border-width;
$caButton-verticalPadding: calc(#{$ca-grid / 2} - #{$caButton-border-width});

.dropdown {
  position: relative;
  padding: $caButton-verticalPadding
    calc(#{$ca-grid * 1} - #{$caButton-border-width});
}

.buttonReset {
  appearance: none;
  display: inline;
  background: transparent;
  color: inherit;
  font: inherit;
  cursor: pointer;
  margin: 0;
  padding: 0;
  border: none;
}

// @TODO - this should be using Kaizen's button
.dropdownButton {
  @include kz-type-inter-button-primary;
  @include ca-inherit-baseline;
  composes: buttonReset;
  cursor: pointer;
  display: flex;
  height: 100%;
  width: 100%;
  align-items: center;
  justify-content: center;
}

.dropdownControlAction {
  @include kz-type-inter-button-primary;
  @include ca-inherit-baseline;
}

.dropdownIcon {
  box-sizing: border-box;
  width: $ca-grid;
  height: $ca-grid;
  padding: 2px;
}

.dropdownIcon + .dropdownLabel {
  margin-left: calc(#{$ca-grid} / 2 - 2px);

  [dir="rtl"] & {
    margin-left: 0;
    margin-right: calc(#{$ca-grid} / 2 - 2px);
  }
}

// If you would like the dropdown icon to only be visible when a certain area is hovered,
// have you hover area use (or compose) the `dropdownHoverArea` class. The label will
// always be visible.
.dropdownHoverArea {
  .dropdownIcon {
    opacity: 0;
    &.isOpen,
    &:focus {
      opacity: 1;
    }
  }
  &:hover .dropdownIcon {
    opacity: 1;
  }
}

.chevronIcon {
  position: relative;
  top: 4px;
}

.menuContainer {
  position: absolute;
  width: $width;
  z-index: $ca-z-index-dropdown;
  right: 0;

  [dir="rtl"] & {
    right: auto;
    left: 0;
  }
}

.reversedColor {
  color: $kz-color-white;
}
