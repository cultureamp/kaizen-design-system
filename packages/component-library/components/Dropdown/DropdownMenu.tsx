import React from "react"

import styles from "./Dropdown.module.scss"

type DropdownMenuProps = {
  hideDropdownMenu: () => void
  position?: {
    top: number
    bottom: number
    left: number
    right: number
  } | null
  children?: React.ReactNode
}

/**
 * @deprecated DropdownMenu is deprecated. Please use draft-menu instead.
 */
class DropdownMenu extends React.Component<DropdownMenuProps> {
  static displayName = "DropdownMenu"
  menu = React.createRef<HTMLDivElement>()

  componentDidMount(): void {
    document.addEventListener("click", this.handleDocumentClick, false)
    window.addEventListener("resize", this.handleDocumentResize, false)
    this.positionMenu()
  }

  componentWillUnmount(): void {
    document.removeEventListener("click", this.handleDocumentClick, false)
    window.removeEventListener("resize", this.handleDocumentResize, false)
  }

  render(): JSX.Element {
    const { hideDropdownMenu, children } = this.props

    return (
      // Disabling instead of addressing because this component is deprecated
      // eslint-disable-next-line jsx-a11y/click-events-have-key-events, jsx-a11y/no-static-element-interactions
      <div
        className={styles.menuContainer}
        ref={this.menu}
        onClick={hideDropdownMenu}
      >
        {children}
      </div>
    )
  }

  positionMenu(): void {
    const menu = this.menu

    if (!this.props.position || !menu) {
      return
    }

    if (menu.current) {
      const pos = this.props.position
      const { innerHeight } = window
      const rect = menu.current.getBoundingClientRect()

      if (pos.bottom > innerHeight - rect.height) {
        menu.current.style.bottom = "24px"
        menu.current.style.top = "auto"
      } else {
        menu.current.style.top = "24px"
        menu.current.style.bottom = "auto"
      }
    }
  }

  handleDocumentClick = (e: MouseEvent): void => {
    if (
      this.menu &&
      this.menu.current &&
      e.target instanceof Node &&
      !this.menu.current.contains(e.target)
    ) {
      this.props.hideDropdownMenu()
    }
  }

  handleDocumentResize = (): void => {
    this.props.hideDropdownMenu()
  }
}

export default DropdownMenu
