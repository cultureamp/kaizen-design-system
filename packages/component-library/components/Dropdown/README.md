# Dropdown


## Status

| Completed | Task                   | Comments                                       |
| --------- | ---------------------- | -----------------------------------------------|
|           | Design Review          |                                                |
|           | Elm Version            | Is in the draft file.                           |
|     ✓     | React Version          |                                                |
|           | Elm Unit Tests         |                                                |
|     ✓     | React Unit Tests       |                                                |
|           | Elm Examples           |                                                |
|     ✓     | React Examples         |                                                |
|           | Cross Browser Testing  |                                                |
|           | Accessibility          |                                                |
|     ✓     | React Styleguide Docs  |http://www.cultureamp.design/components/dropdown|
|           | Elm Styleguide Docs    |                                                |
|           | Flow Type Definitions   |                                                |
|           | Visual Regression Test |                                                |
