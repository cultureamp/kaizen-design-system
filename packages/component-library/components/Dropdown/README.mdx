---
title: Dropdown
navTitle: Dropdown
summaryParagraph: A dropdown is the bit that drops down, and contains links or actions, NOT selected values.
tags: ["Select menu", "Select dropdown", "Multi-select", "Select input"]
needToKnow:
---

## Visuals

### Examples

<iframe style="border: none;" width="744" height="450" src="https://www.figma.com/embed?embed_host=share&url=https%3A%2F%2Fwww.figma.com%2Ffile%2Fe3TyAOZKfLg9cA01Iy1Mgm%2FKaizen-Site-embed%3Fnode-id%3D1319%253A0" allowfullscreen></iframe>

## To keep in mind

*   Dropdowns contain links or actions, not selected values.

## When to use

*   When you’re short on space and need to provide further links or actions.


## When not to use

*   When you’re actually selecting a value from a list of options, use a [Select](/components/select) instead.

## External Links

Here are some examples of other existing design systems:

- [MDN: menu element](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/menu)

