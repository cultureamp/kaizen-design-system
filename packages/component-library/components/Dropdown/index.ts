export { default as Dropdown } from "./Dropdown"
export { DropdownProps } from "./Dropdown"
