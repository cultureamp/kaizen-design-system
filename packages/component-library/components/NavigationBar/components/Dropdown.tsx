import React, { useLayoutEffect, useRef, useState } from "react"

import classNames from "classnames"
import { MenuGroupProps, MenuItemProps } from "../types"
import MenuItem from "./MenuItem"

const styles = require("./Dropdown.module.scss")

type Props = {
  header?: React.ReactElement<any>
  items: Array<MenuItemProps | MenuGroupProps>
}

const Dropdown = ({ items, header }: Props) => {
  const [shiftLeft, setShiftLeft] = useState<boolean>(false)
  const dropdownRef = useRef<HTMLDivElement>(null)

  useLayoutEffect(() => {
    const dropdownElem = dropdownRef.current

    if (dropdownElem) {
      const bounding = dropdownElem.getBoundingClientRect()

      if (
        bounding.right >
        (window.innerWidth || document.documentElement.clientWidth)
      ) {
        // Right side of menu is out of viewport
        setShiftLeft(true)
      }
    }
  }, [])

  const renderMenuGroup = (menuGroup: MenuGroupProps) => {
    return (
      <li className={styles.menuGroup}>
        <h4 className={styles.menuGroupTitle}>{menuGroup.title}</h4>
        {menuGroup.items.map(item => (
          <MenuItem {...item} />
        ))}
      </li>
    )
  }

  return (
    <div
      className={classNames(styles.dropdown, {
        [styles.shiftDropdownLeft]: shiftLeft,
      })}
      ref={dropdownRef}
    >
      {header}
      <ul className={styles.menuItems}>
        {items.map(item => {
          if ("url" in item) {
            return <MenuItem {...item} />
          } else if ("title" in item) {
            return renderMenuGroup(item)
          }
        })}
      </ul>
    </div>
  )
}

export default Dropdown
