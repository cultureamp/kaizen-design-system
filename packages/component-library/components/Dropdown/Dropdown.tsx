import { Icon } from "@kaizen/component-library"
const chevronDownIcon = require("@kaizen/component-library/icons/chevron-down.icon.svg")
  .default
const ellipsisIcon = require("@kaizen/component-library/icons/ellipsis.icon.svg")
  .default
import classNames from "classnames"
import * as React from "react"
import DropdownMenu from "./DropdownMenu"

const styles = require("./Dropdown.module.scss")

type DropdownState = {
  isMenuVisible: boolean
}

export type DropdownProps = {
  icon?: React.SVGAttributes<SVGSymbolElement>
  label?: React.ReactNode
  menuVisible?: boolean
  controlAction?: boolean
  automationId?: string
  reversedColor?: boolean
  iconPosition?: "start" | "end"
}

export default class Dropdown extends React.Component<
  DropdownProps,
  DropdownState
> {
  static displayName = "Dropdown"
  static defaultProps = {
    iconPosition: "start",
  }

  dropdownButton = React.createRef<HTMLButtonElement>()

  constructor(props: DropdownProps) {
    super(props)

    this.state = {
      isMenuVisible: Boolean(props.menuVisible),
    }
  }

  toggleDropdownMenu = (e: React.SyntheticEvent<HTMLButtonElement>) => {
    e.stopPropagation()

    const currentState = this.state.isMenuVisible
    this.setState({
      isMenuVisible: !currentState,
    })
  }

  hideDropdownMenu = () => {
    this.setState({
      isMenuVisible: false,
    })
  }

  getPosition() {
    return this.dropdownButton && this.dropdownButton.current
      ? this.dropdownButton.current.getBoundingClientRect()
      : null
  }

  renderDropdownMenu() {
    return (
      <DropdownMenu
        hideDropdownMenu={this.hideDropdownMenu}
        position={this.getPosition()}
      >
        {this.props.children}
      </DropdownMenu>
    )
  }

  renderIcon = (icon?: React.SVGAttributes<SVGSymbolElement>) => {
    if (!icon) return

    return (
      <span className={styles.dropdownIcon}>
        <Icon icon={icon} role="img" title="Open menu" />
      </span>
    )
  }

  renderDownArrow = () => {
    const { label, controlAction } = this.props
    if (!label || !controlAction) return

    return (
      <span className={styles.chevronIcon}>
        <Icon icon={chevronDownIcon} role="img" title="Open menu" />
      </span>
    )
  }

  renderButtonContent = () => {
    const { label } = this.props
    let { icon } = this.props

    if (!icon && !label) {
      icon = ellipsisIcon
    }

    return (
      <React.Fragment>
        {this.renderIcon(icon)}
        {label && <span className={styles.dropdownLabel}>{label}</span>}
        {this.renderDownArrow()}
      </React.Fragment>
    )
  }

  renderReversedButtonContent = () => {
    const { icon, label } = this.props

    return (
      <React.Fragment>
        {this.renderDownArrow()}
        {label && <span className={styles.dropdownLabel}>{label}</span>}
        {this.renderIcon(icon)}
      </React.Fragment>
    )
  }

  render() {
    const {
      controlAction,
      automationId,
      iconPosition,
      reversedColor,
    } = this.props

    const reverseIcon = iconPosition === "end"
    const btnClass = classNames(styles.dropdownButton, {
      [styles.dropdownControlAction]: controlAction,
      [styles.isOpen]: this.state.isMenuVisible,
      [styles.reversedColor]: reversedColor,
    })
    return (
      <div className={styles.dropdown}>
        <button
          className={btnClass}
          onClick={this.toggleDropdownMenu}
          onMouseDown={e => e.preventDefault()}
          ref={this.dropdownButton}
          data-automation-id={automationId}
        >
          {!reverseIcon && this.renderButtonContent()}
          {reverseIcon && this.renderReversedButtonContent()}
        </button>
        {this.state.isMenuVisible && this.renderDropdownMenu()}
      </div>
    )
  }
}
