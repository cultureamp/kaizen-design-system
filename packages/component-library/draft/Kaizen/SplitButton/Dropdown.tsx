import { Icon } from "@kaizen/component-library"
import * as React from "react"
import { SyntheticEvent } from "react"
import DropdownMenu from "./DropdownMenu"
import { Dir } from "./types"

const chevronDown = require("@kaizen/component-library/icons/chevron-down.icon.svg")
  .default
const styles = require("./styles.scss")

type Props = {
  children: React.ReactNode
  menuVisible?: boolean
  automationId?: string
  dir?: Dir
  dropdownAltText: string
}

type State = {
  isMenuVisible: boolean
}

export default class Dropdown extends React.Component<Props, State> {
  static displayName = "Dropdown"
  static defaultProps = {
    dir: "ltr",
  }
  dropdownButton: HTMLButtonElement | null = null

  constructor(props: Props) {
    super(props)

    this.state = {
      isMenuVisible: Boolean(props.menuVisible),
    }
  }

  toggleDropdownMenu = (e: SyntheticEvent<HTMLButtonElement>) => {
    e.stopPropagation()
    const currentState = this.state.isMenuVisible
    this.setState({
      isMenuVisible: !currentState,
    })
  }

  hideDropdownMenu = () => {
    this.setState({
      isMenuVisible: false,
    })
  }

  getPosition() {
    return this.dropdownButton
      ? this.dropdownButton.getBoundingClientRect()
      : null
  }

  renderDropdownMenu() {
    const { dir } = this.props
    return (
      <DropdownMenu
        hideDropdownMenu={this.hideDropdownMenu}
        position={this.getPosition()}
        dir={dir}
      >
        {this.props.children}
      </DropdownMenu>
    )
  }

  render() {
    const { automationId, dropdownAltText } = this.props
    return (
      <div className={styles.dropdown}>
        <button
          className={styles.dropdownButton}
          onClick={this.toggleDropdownMenu}
          onMouseDown={e => e.preventDefault()}
          ref={k => (this.dropdownButton = k)}
          data-automation-id={automationId}
        >
          <span className={styles.dropdownIcon}>
            <Icon icon={chevronDown} role="img" title={dropdownAltText} />
          </span>
        </button>
        {this.state.isMenuVisible && this.renderDropdownMenu()}
      </div>
    )
  }
}
