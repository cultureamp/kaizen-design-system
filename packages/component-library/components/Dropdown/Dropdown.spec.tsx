import React from "react"
import { render, screen, waitFor } from "@testing-library/react"
import userEvent from "@testing-library/user-event"
import Dropdown from "./Dropdown"

const user = userEvent.setup()

const svgIcon = {
  id: "my-icon",
  viewBox: "0 0 20 20",
}

describe("<Dropdown />", () => {
  it("renders default view", () => {
    const { container } = render(<Dropdown />)

    expect(container.firstChild).toMatchSnapshot()
  })

  it("renders drop down with icon", () => {
    const { container } = render(<Dropdown icon={svgIcon} />)

    expect(container.firstChild).toMatchSnapshot()
  })

  it("renders drop down with icon and label", () => {
    const { container } = render(<Dropdown icon={svgIcon} label="add" />)

    expect(container.firstChild).toMatchSnapshot()
  })

  it("renders drop down with only label", () => {
    const { container } = render(<Dropdown label="add" />)

    expect(container.firstChild).toMatchSnapshot()
  })

  it("renders control action dropdown", () => {
    const { container } = render(
      <Dropdown icon={svgIcon} label="add" controlAction />
    )

    expect(container.firstChild).toMatchSnapshot()
  })

  it("renders reversed color control action dropdown", () => {
    const { container } = render(
      <Dropdown icon={svgIcon} label="add" controlAction reversedColor />
    )

    expect(container.firstChild).toMatchSnapshot()
  })

  it("shows dropdown menu when clicking on the button", async () => {
    const { container } = render(
      <Dropdown>
        <div>Item</div>
      </Dropdown>
    )

    const button = container.querySelector(".dropdownButton")
    button && (await user.click(button))

    await waitFor(() => {
      expect(screen.queryByText("Item")).toBeTruthy()
    })
  })
})
