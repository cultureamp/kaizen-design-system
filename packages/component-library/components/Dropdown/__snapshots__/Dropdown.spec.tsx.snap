// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Dropdown renders control action dropdown 1`] = `
<div
  class="dropdown"
>
  <button
    class="dropdownButton dropdownControlAction"
  >
    <span
      class="dropdownIcon"
    >
      <svg
        class="icon"
        focusable="false"
        role="img"
        viewBox="0 0 20 20"
      >
        <title>
          Open menu
        </title>
        <use
          xlink:href="#my-icon"
        />
      </svg>
    </span>
    <span
      class="dropdownLabel"
    >
      add
    </span>
    <span
      class="chevronIcon"
    >
      <svg
        class="icon"
        focusable="false"
        role="img"
      >
        <title>
          Open menu
        </title>
        <use
          xlink:href="#undefined"
        />
      </svg>
    </span>
  </button>
</div>
`;

exports[`Dropdown renders default view 1`] = `
<div
  class="dropdown"
>
  <button
    class="dropdownButton"
  >
    <span
      class="dropdownIcon"
    >
      <svg
        class="icon"
        focusable="false"
        role="img"
      >
        <title>
          Open menu
        </title>
        <use
          xlink:href="#undefined"
        />
      </svg>
    </span>
  </button>
</div>
`;

exports[`Dropdown renders drop down with icon 1`] = `
<div
  class="dropdown"
>
  <button
    class="dropdownButton"
  >
    <span
      class="dropdownIcon"
    >
      <svg
        class="icon"
        focusable="false"
        role="img"
        viewBox="0 0 20 20"
      >
        <title>
          Open menu
        </title>
        <use
          xlink:href="#my-icon"
        />
      </svg>
    </span>
  </button>
</div>
`;

exports[`Dropdown renders drop down with icon and label 1`] = `
<div
  class="dropdown"
>
  <button
    class="dropdownButton"
  >
    <span
      class="dropdownIcon"
    >
      <svg
        class="icon"
        focusable="false"
        role="img"
        viewBox="0 0 20 20"
      >
        <title>
          Open menu
        </title>
        <use
          xlink:href="#my-icon"
        />
      </svg>
    </span>
    <span
      class="dropdownLabel"
    >
      add
    </span>
  </button>
</div>
`;

exports[`Dropdown renders drop down with only label 1`] = `
<div
  class="dropdown"
>
  <button
    class="dropdownButton"
  >
    <span
      class="dropdownLabel"
    >
      add
    </span>
  </button>
</div>
`;

exports[`Dropdown renders reversed color control action dropdown 1`] = `
<div
  class="dropdown"
>
  <button
    class="dropdownButton dropdownControlAction reversedColor"
  >
    <span
      class="dropdownIcon"
    >
      <svg
        class="icon"
        focusable="false"
        role="img"
        viewBox="0 0 20 20"
      >
        <title>
          Open menu
        </title>
        <use
          xlink:href="#my-icon"
        />
      </svg>
    </span>
    <span
      class="dropdownLabel"
    >
      add
    </span>
    <span
      class="chevronIcon"
    >
      <svg
        class="icon"
        focusable="false"
        role="img"
      >
        <title>
          Open menu
        </title>
        <use
          xlink:href="#undefined"
        />
      </svg>
    </span>
  </button>
</div>
`;
